File: clients/rs/src/features/config_embed_wrap_package.rs
use std::{error::Error, path::Path, fs::{self}, sync::Arc};
use polywrap_client::{wasm::wasm_package::WasmPackage, core::{file_reader::SimpleFileReader, invoker::Invoker}, client::PolywrapClient, builder::{PolywrapClientConfigBuilder, PolywrapClientConfig}};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::input::{expect_root_dir};

#[derive(Deserialize, Serialize)]
struct Args {
  first: u8,
  second: u8
}

#[derive(Deserialize)]
struct InputObj {
  directory: String,
  method: String,
  args: Args
}

pub fn run_test_case(input: &Value) -> Result<(), Box<dyn Error>> {
  let input_obj: InputObj = serde_json::from_value(input.clone())?;

  let wrap_dir = expect_root_dir(
      &input_obj.directory,
      std::env::current_dir()?.join("../../").to_str().unwrap(),
  )?;

  let method = input_obj.method;
  let args = input_obj.args;

  println!("Reading wrap.info & wrap.wasm from {}", input_obj.directory.to_string());

  let manifest = fs::read(Path::new(&wrap_dir).join("wrap.info"))?;
  let wasm_module = fs::read(Path::new(&wrap_dir).join("wrap.wasm"))?;

  println!("Creating WrapPackage from raw wrap.info & wrap.wasm bytes");

  let wrap_package = WasmPackage::from_bytecode(wasm_module, Arc::new(SimpleFileReader::new()), Some(manifest));

  println!("Adding WrapPackage to ClientConfig");

  let mut config = PolywrapClientConfig::new();
  config.add_package("embed/foo".try_into().unwrap(), Arc::new(wrap_package));
  
  let client: PolywrapClient = PolywrapClient::new(config.into());

  println!("Invoking WrapPackage");

  let result = client.invoke_raw(
    &"embed/foo".try_into().unwrap(),
    &method,
    Some(&polywrap_client::msgpack::to_vec(&args)?),
    None,
    None
  );

  if result.is_ok() {
    println!("Success!");
  }

  Ok(())
}

File: clients/swift/Sources/Readiness/Features/ConfigEmbedWrapPackage.swift
import Foundation
import PolywrapClient

public struct u8MethodArgs: Codable {
    var first: Int
    var second: Int
    
    public init(first: Int, second: Int) {
        self.first = first
        self.second = second
    }
}

struct ConfigEmbedWrapPackageTest: Feature {
   func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any],
            let directory = inputObj["directory"] as? String else {
           throw InputError.expectedObject
        }


        print("Reading wrap.info & wrap.wasm from \(directory)")

        let wrapDir = try Input.expectRootDir(directory, rootDir: "./")
        let method = try Input.expectString(inputObj["method"]!)

        guard let argsDict = inputObj["args"] as? [AnyHashable: Any],
            let first = argsDict["first"] as? Int,
            let second = argsDict["second"] as? Int else {
            throw InputError.expectedObject
        }

        let args = u8MethodArgs(first: first, second: second)
        let reader = ResourceReader(bundle: Bundle.module)
        let module = try reader.readFile(wrapDir)

        print("Creating WrapPackage from raw wrap.info & wrap.wasm bytes")

        let package = WasmPackage(reader: nil, module: module)
        print("Adding WrapPackage to ClientConfig")

        let uri = try! Uri("embed/foo")

        let builder = BuilderConfig().addPackage(uri, package)

        let client = builder.build()
        print("Invoking WrapPackage")
        let result: Int? = try? client.invoke(uri: uri, method: method, args: args)

        if result != nil {
            print("Success!")
        }
   }
}
